src/sentry/api/endpoints/artifact_lookup.py

```diff
@@ -0,0 +1,336 @@
+        # the file names we are querying for, and also leave us with the remaining
+        # set of file names that are not covered by any bundle, to look up below
+
+        bundle_file_ids = collect_artifact_bundles_containing_debug_ids(debug_ids, project)
+        individual_files = try_resolve_urls(urls, project, release_name, dist_name, bundle_file_ids)
+
+        # Then: Construct our response
+        url_constructor = UrlConstructor(request)
+
+        found_artifacts = []
+        for file_id in bundle_file_ids:
+            found_artifacts.append(
+                {
+                    "type": "bundle",
+                    "url": url_constructor.url_for_file_id(file_id),
+                }
+            )
+
+        for file in individual_files:
+            found_artifacts.append(
+                {
+                    "type": "file",
+                    "url": url_constructor.url_for_file_id(file.id),
+                    # The `name` is the url/abs_path of the file,
+                    # as in: `"~/path/to/file.min.js"`.
+                    "abs_path": file.name,
+                    # These headers should ideally include the `Sourcemap` reference
+                    "headers": file.headers,
+                }
+            )
+
+        # NOTE: We do not paginate this response, as we have very tight limits
+        # on all the individual queries.
+        return Response(serialize(found_artifacts, request.user))
+
+
+def collect_artifact_bundles_containing_debug_ids(
+    debug_ids: List[str], project: Project
+) -> Set[int]:
+    # For debug_ids, we will query for the artifact_bundle/file_id directly
+    bundle_file_ids = set(
+        DebugIdArtifactBundle.objects.filter(
+            organization_id=project.organization.id,
+            debug_id__in=debug_ids,
+        )
+        .select_related("artifact_bundle")
+        .values_list("artifact_bundle__file_id", flat=True)
+        .distinct("artifact_bundle__file_id")[: MAX_BUNDLES_BY_DEBUG_ID + 1]
+    )
+
+    if len(bundle_file_ids) > MAX_BUNDLES_BY_DEBUG_ID:
+        logger.error(
+            "querying for artifact bundles by `debug_id` yielded more than %s results",
+            MAX_BUNDLES_BY_DEBUG_ID,
+        )
+
+    return bundle_file_ids
+
+
+def try_resolve_urls(
+    urls: List[str], project: Project, release_name: str, dist_name: str, bundle_file_ids: Set[int]
+) -> Sequence[File]:
+    if not urls:
+        return list()
+
+    if release_name is None:
+        logger.error("trying to look up artifacts by `url` without a `release`")
+        return list()
+
+    # If we have `urls`, we want to:
+    # First, get the newest X artifact bundles, and *look inside them*
+    # to figure out if the file is included in any of them
+    urls = collect_release_artifact_bundles_containing_urls(
+        urls, project, release_name, dist_name, bundle_file_ids
+    )
+    if not urls:
+        return list()
```

Comment:
What confuses me about this implementation is that, coupling the shared state of `bundle_file_ids` with the naming of `urls` it took me 5 minutes to understand properly the logic.

I think that just renaming the variable would help a lot to convey the variable semantics.
```suggestion
    if not remaning_urls:
        return list()
```